crawler/jobs/cherrypicker_test_data.py

```diff
@@ -0,0 +1,154 @@
+from bson.objectid import ObjectId
+from datetime import datetime
+from functools import reduce
+import json
+import logging
+import os
+from typing import List
+
+from crawler.constants import (
+    COLLECTION_CHERRYPICK_TEST_DATA,
+    FIELD_MONGODB_ID,
+    FIELD_UPDATED_AT,
+    FIELD_STATUS,
+    FIELD_PLATE_SPECS,
+    FIELD_BARCODES,
+    FIELD_FAILURE_REASON,
+    FIELD_STATUS_PENDING,
+    FIELD_STATUS_STARTED,
+    FIELD_STATUS_PREPARING_DATA,
+    FIELD_STATUS_CRAWLING_DATA,
+    FIELD_STATUS_COMPLETED,
+    FIELD_STATUS_FAILED,
+    TEST_DATA_CENTRE_PREFIX,
+    TEST_DATA_ERROR_NO_RUN_FOR_ID,
+    TEST_DATA_ERROR_WRONG_STATE,
+    TEST_DATA_ERROR_INVALID_PLATE_SPECS,
+    TEST_DATA_ERROR_NUMBER_OF_PLATES,
+    TEST_DATA_ERROR_NUMBER_OF_POS_SAMPLES,
+)
+from crawler.db.mongo import (
+    create_mongo_client,
+    get_mongo_collection,
+    get_mongo_db,
+)
+from crawler.helpers.cherrypicker_test_data import (
+    create_barcodes,
+    create_barcode_meta,
+    create_csv_rows,
+    write_plates_file,
+)
+from crawler.helpers.general_helpers import get_config
+from crawler.main import run as run_crawler
+from crawler.types import Config
+
+
+logger = logging.getLogger(__name__)
+
+
+class TestDataError(Exception):
+    def __init__(self, message):
+        self.message = message
+
+
+def process(run_id: str, config: Config = None) -> List[List[str]]:
+    """Generates cherrypicker test data for processing by Crawler and then
+    processes it via the usual runner.
+
+    The specification of the plates to be generated should be in Mongo. Each
+    plate will contain an exact number of positive results between 0 and 96 as
+    specified. Up to 100 plates can be generated at a time.
+
+    Arguments:
+        run_id: str - The ID of the run.  If this is not found in Mongo an
+            exception will be thrown.
+
+    Returns:
+        Metadata about the plates generated, as:
+        [ [ "barcode1", "description1" ], [ "barcode2", "description2" ] ]
+    """
+    logger.info("Begin generating data.")
+
+    if config is None:
```

Comment:
quite a long method with lots of nested with if try clauses, possible to refactor?
appreciate some of the other crawler code is much worse on this!

Reply:
I was thinking the same thing as I was testing it.  It actually got worse during testing because of the edge cases I realised weren't catered for.  I'll do a refactor now.